instead of yml mapping use java based, added jwt filter

mkhomishchak opened the pull request:
moved gateway config to java class as only it allows to add custom filters implemented jwt filter that will use auth-service for token validation /crypto-pricing will be secured later once we will implement this part in service itself first

mkhomishchak committed:
instead of yml mapping use java based, added jwt filter

moved gateway config to java class as only it allows to add custom filters
implemented jwt filter that will use auth-service for token validation
/crypto-pricing will be secured later once we will implement this part in service itself first

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/GatewayConfig.java:
```diff
@@ -0,0 +1,75 @@
+    @Value("${AUTHENTICATION_SERVICE_URL:http://localhost:8088}")
+    private String authenticationServiceUrl;
+
+    @Value("${EXCHANGER_SERVICE_URL:http://localhost:8085}")
+    private String exchangerServiceUrl;
+
+    @Value("${FEEDBACK_SERVICE_URL:http://localhost:8085}")
+    private String feedbackServiceUrl;
+
+    @Value("${GOAL_SERVICE_URL:http://localhost:8085}")
+    private String goalServiceUrl;
+
+    @Value("${USERGOAL_SERVICE_URL:http://localhost:8085}")
+    private String userServiceUrl;
```
Are u sure that u want to set as default port for 4 services as 8085?

mkhomishchak replied:
currently all of those services are in monorepo, once they will be moved to separate service I will change the port

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,74 @@
+
+    private void updateRequest(ServerWebExchange exchange, Long userId){
+        exchange.getRequest().mutate()
+                .header("UserId", String.valueOf(userId))
```
Move UserId in a constant

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/filters/AuthenticationFilter.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private String getAuthHeader(ServerHttpRequest request) {
+        return request.getHeaders().getOrEmpty(HttpHeaders.AUTHORIZATION).get(0);
```
You are using get after getOrEmpty not ensuring that the list of values is not empty. The attempt to use get with index in empty list will throw indexoutofbound.

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/service/JwtServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public ProcessedTokenResp processToken(String token) {
```
Token might be null.

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/service/JwtServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+        return new HttpEntity<>(headers);
+    }
+
+    private HttpHeaders createHttpRequestHeaders(Map<String, String> headerParams) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        headerParams.forEach(headers::set);
+        return headers;
```
1. If your headerParams can be nullable, then okay. If not - provide a check for it.
2. Instead of iterating over headerParams with forEach and setting headers individually, you can use the addAll method which is provided by HttpHeaders. This simplifies the code and improves readability.

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/validator/RouterValidator.java:
```diff
@@ -0,0 +1,19 @@
+@Component
+public class RouterValidator {
+
+    public static final List<String> openApiEndpoints = List.of(
```
I would recommend u to use unmodifiable list here to ensure that the list remains unmodifiable, even if someone tries to modify it later unintentionally:

```
public static final List<String> openApiEndpoints = Collections.unmodifiableList(
        List.of("/auth/register", "/auth/login", "/users", "/users/authenticate", "/exchangers/used-currencies")
);
```

mkhomishchak replied:
List.of() creates immutable List as well

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/validator/RouterValidator.java:
```diff
@@ -0,0 +1,19 @@
+            "/auth/register", "/auth/login", "/users", "/users/authenticate", "/exchangers/used-currencies"
+    );
+
+    public Predicate<ServerHttpRequest> isSecured =
+            request -> openApiEndpoints.stream()
+                    .noneMatch(uri -> request.getURI().getPath().contains(uri));
```
The method name isSecured might be a bit misleading because it seems to check if the endpoint is not in the list of open API endpoints. I would recommend u to renaming it to something like isPublicEndpoint to better convey its intent BLYAT

COCONUTmkII reviewed (changes requested):
Overall looks good, but you can improve something nahyi blyat. Be ensure that your microservices works correctly with your gateway.

mkhomishchak committed:
more error handling, improve naming

added more error handling for null and not present scenarios
improved naming to indicate exectly what method is supposed to do

COCONUTmkII commented on src/main/java/com/khomishchak/apigateway/service/JwtServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+    private HttpHeaders createHttpRequestHeaders(Map<String, String> headerParams) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        MultiValueMap<String, String> multiValueMap = new LinkedMultiValueMap<>();
```
Why u created this map?

mkhomishchak replied:
as addAll() of HttpHeaders requries it to be provided, it allows us to have list of values for single header what can be usefull in the future

COCONUTmkII reviewed (approved):
I added a comment in one method because I don't understand the map creation. Maybe u made a mistake. If u can explain in details - okay. Approved for now.
